### PR TITLE
[feature] Updated code to handle ECDSA signature algorithm #118

### LINKED ISSUE #118: Importing existing CA: KeyError: 'ecdsa-with-SHA384'
I'm trying to importing an existing CA certificate which generates an error:

```
  File "/home/max/git/django-x509/django_x509/base/models.py", line 362, in _import
    self.digest = SIGNATURE_MAPPING[algorithm]
KeyError: 'ecdsa-with-SHA384'
[13/Jun/2021 00:03:38] "POST /admin/django_x509/ca/add/ HTTP/1.1" 500 142363
```

Looking into `django_x509/base/models.py` dict `SIGNATURE_MAPPING`: is are EC signatures and certificates/keys supported?

### PR DESCRIPTION
**Pull request by praptisharma28**
Fixes #118

### CODE REVIEW

**Review by nemesifier — CHANGES_REQUESTED** [review]
I did a bit of research. Since we are adding this algorithm, it doesn't make much sense to not add other algorithms too.

Since we are at it, let's patch it so that more users will be able to use this library.

Until today we have worked mostly with RSA, but other type of algorithm are gaining momentum and would be good to support them.

## x509 algorithms

Python libraries for handling X.509 certificates typically support a wide range of algorithms for signing and verifying certificates. These algorithms include various RSA, ECDSA, and SHA-based options. Here are some common algorithms supported by popular Python libraries such as cryptography, pyOpenSSL, and M2Crypto:

### Common Algorithms

**RSA-based Algorithms:**

- sha1WithRSAEncryption
- sha256WithRSAEncryption
- sha384WithRSAEncryption
- sha512WithRSAEncryption
- ~md5WithRSAEncryption~ (not recommended due to security vulnerabilities)

**ECDSA-based Algorithms:**

- ecdsa-with-SHA1
- ecdsa-with-SHA256
- ecdsa-with-SHA384
- ecdsa-with-SHA512

**DSA-based Algorithms:**

- dsaWithSHA1
- dsaWithSHA256

**EdDSA Algorithms:**

- Ed25519
- Ed448

Can you please add these as well?

**Comment by praptisharma28** [comment]
> I did a bit of research. Since we are adding this algorithm, it doesn't make much sense to not add other algorithms too.
> 
> Since we are at it, let's patch it so that more users will be able to use this library.
> 
> Until today we have worked mostly with RSA, but other type of algorithm are gaining momentum and would be good to support them.
> 
> ## x509 algorithms
> Python libraries for handling X.509 certificates typically support a wide range of algorithms for signing and verifying certificates. These algorithms include various RSA, ECDSA, and SHA-based options. Here are some common algorithms supported by popular Python libraries such as cryptography, pyOpenSSL, and M2Crypto:
> 
> ### Common Algorithms
> **RSA-based Algorithms:**
> 
> * sha1WithRSAEncryption
> * sha256WithRSAEncryption
> * sha384WithRSAEncryption
> * sha512WithRSAEncryption
> * ~md5WithRSAEncryption~ (not recommended due to security vulnerabilities)
> 
> **ECDSA-based Algorithms:**
> 
> * ecdsa-with-SHA1
> * ecdsa-with-SHA256
> * ecdsa-with-SHA384
> * ecdsa-with-SHA512
> 
> **DSA-based Algorithms:**
> 
> * dsaWithSHA1
> * dsaWithSHA256
> 
> **EdDSA Algorithms:**
> 
> * Ed25519
> * Ed448
> 
> Can you please add these as well?

Yes @nemesifier will work upon it.

File: django_x509/base/models.py [hunk 1]
```diff
@@ -32,6 +32,14 @@
...
     ('sha384', 'SHA384'),
     ('sha512', 'SHA512'),
+    ('ecdsa-with-sha1', 'ECDSA with SHA1'),
+    ('ecdsa-with-sha256', 'ECDSA with SHA256'),
+    ('ecdsa-with-sha384', 'ECDSA with SHA384'),
+    ('ecdsa-with-sha512', 'ECDSA with SHA512'),
+    ('dsaWithSHA1', 'DSA with SHA1'),
+    ('dsaWithSHA256', 'DSA with SHA256'),
+    ('ed25519', 'Ed25519'),
+    ('ed448', 'Ed448'),
```
Comment (nemesifier) [inline django_x509/base/models.py:42]: I don't think it makes sense to alter the digest choices. Try creating a new certificate or CA with any of these and watch it fail.

**Review by nemesifier — CHANGES_REQUESTED** [review]
The suggestion I have given in https://github.com/openwisp/django-x509/issues/118#issuecomment-860123105 is incorrect.

I assumed we just had to change that part of the code to support generating certificates with other algorithms but that's not the case.

The digest options being added here are invalid, try to create a CA or a cert from the admin interface with any of those and you'll see it doesn't work.

To support this feature properly, we must add a way to specify the algorithm used for generating the certificate, which now is hardcoded to be `crypto.TYPE_RSA` (RSA).

**Comment by praptisharma28** [comment]
> The suggestion I have given in [#118 (comment)](https://github.com/openwisp/django-x509/issues/118#issuecomment-860123105) is incorrect.
> 
> I assumed we just had to change that part of the code to support generating certificates with other algorithms but that's not the case.
> 
> The digest options being added here are invalid, try to create a CA or a cert from the admin interface with any of those and you'll see it doesn't work.
> 
> To support this feature properly, we must add a way to specify the algorithm used for generating the certificate, which now is hardcoded to be `crypto.TYPE_RSA` (RSA).
Okay

**Comment by praptisharma28** [comment]
> > The suggestion I have given in [#118 (comment)](https://github.com/openwisp/django-x509/issues/118#issuecomment-860123105) is incorrect.
> > I assumed we just had to change that part of the code to support generating certificates with other algorithms but that's not the case.
> > The digest options being added here are invalid, try to create a CA or a cert from the admin interface with any of those and you'll see it doesn't work.
> > To support this feature properly, we must add a way to specify the algorithm used for generating the certificate, which now is hardcoded to be `crypto.TYPE_RSA` (RSA).


Okay

**Comment by nemesifier** [comment]
Will close this for now as I don't think we'll work on it.

**Pull request closed without merging**